### DISCUSSION TITLE
Spec Exploration for `meltano catalog` UX

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/meltano#6766] [General]
Following from

- https://github.com/meltano/meltano/discussions/6757

I wanted to explore what a `meltano catalog` CLI command group might look like from a UX perspective.

Two fundamental use cases as I can see:

## Use case 1: Maintaining behind-the-scenes internal catalog artifacts

E.g.:

- `meltano catalog rm tap-gitlab` - removes the internally cached `catalog.json` artifact for the extractor
- `meltano catalog update tap-gitlab` - explicitly updates the cached `catalog.json` artifact for the named extractor

## Use case 2: Printing catalog info to STDOUT

E.g.:

- `meltano catalog select --list` (Presumably the same `meltano select` functionality but moved under the `catalog` command group.

## Use case 2: Generating and/or maintaining one or more source-controlled "data catalog" artifacts

This is the area I want to explore more. We don't have any source-controlled data catalog artifacts as of now, but one possible piece of prior art would be the automated update of `dbt`'s `source.yml` file based on a scan of the target.

Probably we will want to explicitly create a **catalog** and then add data assets to the catalog. If we map the things added to specific state IDs or to specific State ID naming patterns, then we can have a fairly automated approach to maintaining them.

```console
// Create a new catalog artifact named `raw_github`:
> meltano catalog new raw_github --type=dbt
// Register a couple specific State IDs as affecting the `raw_github` metadata artifact:
> meltano catalog append raw_github --state-id=dev:tap-github-comments-to-target-snowfake
> meltano catalog append raw_github --state-id=dev:tap-github-search-to-target-snowfake
// Sync the catalog based on available data for the named state ID:
> meltano catalog sync raw_github

// Create a new catalog artifact called `snowflake_dev`
> meltano catalog new snowflake_dev --type=dbt
// Register all matching state IDs as affecting the `raw_github` metadata artifact
> meltano catalog append --catalog=snowflake_dev --state_id_pattern=dev:*-target-snowfake

// Run the EL and auto-sync the catalog.
// Will create a diff in the catalog file _only_ if anything has changed since last run:
> meltano run --sync-catalog=true tap-github target-snowflake
// Optionally send a custom alert after sync if git status is dirty for catalog files in the repo.
```

This breaks down a few different components:

1. Catalog **caching** is an _internal_ process we do today. Specifically, we cache `catalog.json` files internally and we apply selection rules to the `catalog.json` file in order to govern tap selection behaviors.
2. Catalog **creation** initializes a specific 'catalog' entity, instantiating the 'catalog' as a source-controlled artifact.
3. Catalog **association** links the catalog artifact to a specific set of EL pairs (proposed above as mapped to state IDs or state ID naming patterns).
4. Catalog **syncing** is the process of updating the project artifact based on content in the cache. (Where the cache can be generated on-the-fly or be preexisting from a prior run).

cc @tayloramurphy

**Comment by tayloramurphy** [comment]
I've made a milestone to gather many of the catalog related issues: https://github.com/meltano/meltano/milestone/12

With our refocus on EL it's going to be more important that we get the experience around the catalog, state, and mappers right as they're "must haves" and competitive advantages. 

https://github.com/meltano/meltano/issues/3421 in particular is a more recent issue highlighting the new command.

**Reply by aaronsteers** [reply]
Totally agreed that this isn't a high priority as of now. It comes up often enough though, that I wanted to get some thoughts down in terms of how we might eventually deliver.

Specifically: the idea of creating a named catalog _by request_ and running the `sync` command to make changes to the created file(s) on demand.

**Reply by tayloramurphy** [reply]
> Totally agreed that this isn't a high priority as of now

@aaronsteers actually I meant to imply that this will become a high priority in relatively short order. 😅  While we work on Managed there are key areas of Meltano OSS that should be improved - namely catalog, state, and mapper support. Our current focus is on state (Cody's work) and I'll be bringing forward the catalog and mapper work post-tap-toberfest 😄

**Reply by aaronsteers** [reply]
@tayloramurphy - Oh nice. Yeah, thanks for clarifying. 👍😁

**Comment by aaronsteers** [comment]
## Re: Creating our own Meltano-specific catalog syntax vs adopting dbt's syntax or another provider's syntax

I think a hard part of of this is whether to create our own catalog metadata format, versus adopting a 3rd party's catalog metadata format.

One good option is to just decide we may want to have catalogs of different types, and to support a `--type` argument in `meltano catalog new` (or `create`, etc.). 

This opens up a very expedient path forward, which is to _first_ add `--type=dbt` syntax, which would be represented by a `source.yml` syntax. What's nice about this approach is that it immediately renders in `dbt docs`, and it _also_ immediately solves the problem of users having to register their datasets in dbt before being able to transform them...

**Reply by aaronsteers** [reply]
A very simple proposal for a Meltano-specific catalog syntax emerged in

- https://github.com/meltano/meltano/issues/6876#issuecomment-1279899714,

 which proposes `streams` cataloged with the properties `name` (aka 'stream name'), `schema` (the JSON Schema of _selected_ streams and properties), `key-properties`, and `replication-key`. Otherwise we just would need to decide on a spelling for each that we can live with long-term: `key-properties`/`key_properties`/`primary_keys` and `replication_key`/`replication-key`/`incremental-replication-key`, etc. If we stick with Singer precedent, I believe the spellings would be `key-properties` and `replication-key`.

Note:

- As of now, it's problematic to use `replication-method` here because we don't dictate that directly and we can't directly observe it on a per stream level. Many taps may choose the replication method directly or indirectly based on a combination of the user 'config' preferences and each steams' available replication options.

**Comment by aaronsteers** [comment]
Related discussion here:

- https://github.com/meltano/meltano/issues/6876#issuecomment-1279899714
